### PR TITLE
StructuredCVector3

### PR DESCRIPTION
**Pull request by mancoast**
screenshot from 2016-09-25 20-58-14

### CODE REVIEW

File: mandelbulber2/src/algebra.hpp [hunk 1]
```diff
@@ -240,6 +240,38 @@ inline CVector3 fabs(CVector3 v)
...
+			x[n] = vectors[n].x;
+			y[n] = vectors[n].y;
+			z[n] = vectors[n].z;
+		}
+	}
+
+	double x[8];
+	double y[8];
+	double z[8];
+};
```
Comment (mancoast) [inline mandelbulber2/src/algebra.hpp]: Structure of Arrays enables SIMD

File: mandelbulber2/src/algebra.hpp [hunk 2]
```diff
@@ -240,6 +240,38 @@ inline CVector3 fabs(CVector3 v)
...
+	inline StructuredCVector3()
+	{
+		for (int n = 0; n < 8; n++)
+		{
+			x[n] = 0.0;
+			y[n] = 0.0;
+			z[n] = 0.0;
+		}
+	}
+
```
Comment (mancoast) [inline mandelbulber2/src/algebra.hpp]: Structure is initialized using array of 8 CVector3's

File: mandelbulber2/src/compute_fractal.cpp [hunk 3]
```diff
@@ -110,10 +110,12 @@ void Compute(const cNineFractals &fractals, const sFractalIn &in, sFractalOut *o
...
+	StructuredCVector3 lastGoodZ;
+	StructuredCVector3 lastZ;
 
-	for (i = 0; i < in.maxN; i++)
+	// chop this loop into sets of 8
+	//   to use StructuredCVector3 and double*
+	for (i = 0; i < in.maxN; i = i + 8)
 	{
 		lastGoodZ = lastZ;
 		lastZ = z;
```
Comment (mancoast) [inline mandelbulber2/src/compute_fractal.cpp]: Can these loop iterations be subdivided into sets of 8?
The use of `lastGoodZ = lastZ; lastZ = z; `makes me nervous about Loop dependence analysis.
Can we safely execute 8 initial iterations in paralel and step through this loop with increments of 8?
`for (i = 0; i < in.maxN; i = i + 8)`
What are you thoughts?

**Comment by buddhi1980** [comment]
We cannot execute this loop in parallel because initial value for each
iteration (z variable) is taken from previous iteration. This loop cannot
be parallelized.

26 wrz 2016 01:57 "mancoast" <notifications@github.com> napisał(a):

File: mandelbulber2/src/shaders.cpp [hunk 4]
```diff
@@ -730,6 +730,7 @@ sRGBAfloat cRenderWorker::AmbientOcclusion(const sShaderInputData &input)
 		double opacity = 0.0;
 		double shadowTemp = 1.0;
```
Comment (mancoast) [inline mandelbulber2/src/shaders.cpp]: I want to slice these loop iterations into strides to exploit vectors.

screenshot from 2016-09-29 16-45-10
screenshot from 2016-09-29 16-44-09

**Comment by buddhi1980** [comment]
Instead of trying to optimize this part of code, which would be very difficult for SIMD, try to look at CMatrix33 or CRotationMatrix class, which could be better for that. They are in algebra.hpp.

**Comment by mancoast** [comment]
I will investigate CMatrix33 and CRotationMatrix for some initial implementations in the codebase.

For the application to truely benefit from SIMD, it needs to be implemented within the loops that retire majority instructions.

**Comment by mancoast** [comment]
I suggest we examine usage of 

http://www.agner.org/optimize/vectorclass.pdf

```
"Explain this
with an example:
// Example 1a. Calculations on arrays
float a[8], b[8], c[8]; // declare arrays
... // put values into arrays
for (int i = 0; i < 8; i++) { // loop for 8 elements
 c[i] = a[i] + b[i]*1.5f; // operations on each element
}
The vector class library allows you to write this code as vectors:
// Example 1b. Same code using vectors
#include "vectorclass.h" // use vector class library
float a[8], b[8], c[8]; // declare arrays
... // put values into arrays
Vec8f avec, bvec, cvec; // define vectors
avec.load(a); // load array a into vector
bvec.load(b); // load array b into vector
cvec = avec + bvec * 1.5f; // do operations on vectors
cvec.store(c); // save result in array c

Example 1b does the same as example 1a, but more efficiently because it
utilizes SIMD instructions that do eight additions and/or eight multiplications in a
single instruction. Modern microprocessors have these instructions which may
give you a throughput of eight floating point additions and eight multiplications
per clock cycle. A good optimizing compiler may actually convert example 1a
automatically to use the SIMD instructions, but in more complicated cases you
cannot be sure that the compiler is able to vectorize your code in an optimal way"

```

**Pull request merged**